Text-to-SQL Project.

Study of one of the IBM Granite model, here the 20B parameters code base 8K, for the generation of SQL queries from question in natural language.

Partie 1.

Load the model and the dataset

In [6]:
!pip install transformers bitsandbytes accelerate datasets outlines scikit-learn sqlvalidator

In [41]:


from transformers import (AutoModelForCausalLM,BitsAndBytesConfig, AutoTokenizer)

import torch


hf_token = "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"
model_path = "ibm-granite/granite-20b-code-base-8k"
#Load the tokenizer associated on the model
tokenizer = AutoTokenizer.from_pretrained(model_path, token=hf_token)
# Prevent some transformers specific issues.
tokenizer.use_default_system_prompt = False
tokenizer.pad_token_id = tokenizer.eos_token_id

# We want to use 4bit quantization to save memory
quantization_config = BitsAndBytesConfig(
    load_in_8bit=False, load_in_4bit=True
)

#Load the model
model = AutoModelForCausalLM.from_pretrained(model_path,quantization_config=quantization_config,
                                             device_map={"": 0},
                                             torch_dtype=torch.bfloat16,
                                             token=hf_token)
#Set the model for evaluation
model.eval()





The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

tokenizer_config.json:   0%|          | 0.00/4.11k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.06M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/1.02k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/644 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/47.0k [00:00<?, ?B/s]

model-00001-of-00009.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

model-00002-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00003-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00004-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00005-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00006-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00007-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00008-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00009-of-00009.safetensors:   0%|          | 0.00/680M [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/9 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/132 [00:00<?, ?B/s]

GPTBigCodeForCausalLM(
  (transformer): GPTBigCodeModel(
    (wte): Embedding(49152, 6144)
    (wpe): Embedding(8192, 6144)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-51): 52 x GPTBigCodeBlock(
        (ln_1): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
        (attn): GPTBigCodeSdpaAttention(
          (c_attn): Linear4bit(in_features=6144, out_features=6400, bias=True)
          (c_proj): Linear4bit(in_features=6144, out_features=6144, bias=True)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
        (mlp): GPTBigCodeMLP(
          (c_fc): Linear4bit(in_features=6144, out_features=24576, bias=True)
          (c_proj): Linear4bit(in_features=24576, out_features=6144, bias=True)
          (act): GELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): Laye

In [105]:
from transformers import GenerationConfig

generation_config = GenerationConfig(
  max_new_tokens = 1000,
  min_new_tokens = 10,
  do_sample=False,
  temperature=0.3,
  repetition_penalty=1.0,
  eos_token_id=tokenizer.eos_token_id,
  pad_token_id=tokenizer.pad_token_id,
)

/usr/local/lib/python3.10/dist-packages/transformers/generation/configuration_utils.py:628: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.3` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.
  warnings.warn(


In [106]:
from datasets import load_dataset

#Dataset: https://huggingface.co/datasets/xlangai/spider?row=17
ds = load_dataset("xlangai/spider")

print(ds)



DatasetDict({
    train: Dataset({
        features: ['db_id', 'query', 'question', 'query_toks', 'query_toks_no_value', 'question_toks'],
        num_rows: 7000
    })
    validation: Dataset({
        features: ['db_id', 'query', 'question', 'query_toks', 'query_toks_no_value', 'question_toks'],
        num_rows: 1034
    })
})


In [178]:
#Only looking for the validation test of the dataset
examples = [
    {"question": ex["question"], "SQL_query": ex["query"], "db_id": ex["db_id"]}
    for ex in ds['validation'].to_list()
]

#print(examples[0])

# Split examples and keep only a few samples to have short computation time.
test, train = examples[:100], examples[100:500]
print(f"Train len {len(train)}. Test len {len(test)}")
#print(f"Validation len {len(validation_dataset)}")
print(f"First example of test:\n{test[0]}")


Train len 400. Test len 100
First example of test:
{'question': 'How many singers do we have?', 'SQL_query': 'SELECT count(*) FROM singer', 'db_id': 'concert_singer'}


Partie 2 : Génération and evaluation function

In [120]:
def generate(prompt, llm=model, generation_config=generation_config):

  # Create turns with the given prompt

  # Tokenize turns, tokenize the text
  input_tokens = tokenizer(prompt, return_tensors='pt').to('cuda')
  input_ids=input_tokens['input_ids']

  # Ensure we don't use gradient to save memory space and computation time.
  #Generate output tokens
  with torch.no_grad():
    outputs = llm.generate(input_ids).to('cuda')

  #decode output token into text

  # Recover and decode answer.
  answer_tokens = outputs[0, input_ids.shape[1]:-1]

  #decode output token into text
  #.strip() delete space useless

  answer = tokenizer.decode(answer_tokens).strip()  # Remove unnecessary spaces

  # Remove the "ASSISTANT: " prefix, if present
  if answer.startswith("ASSISTANT:"):
    answer = answer[len("ASSISTANT:"):].strip()  # Remove the prefix and strip spaces

  return answer


In [170]:
import re

def extract_sql_answer(answer: str):
    # Recherche une ligne qui commence par "SELECT", "INSERT", "UPDATE", ou d'autres mots-clés SQL
    text=answer.replace('\n', ' ').strip()
    match = re.search(r"(SELECT|INSERT|UPDATE|DELETE|CREATE|DROP|ALTER).*", text.strip(), re.IGNORECASE)
    if match:
        query = match.group(0).strip()  # Retourne la première requête SQL trouvée
        query = re.sub(r'<<<.*$', '', query).strip()
        return query
    else:
        return text



In [7]:
import sqlparse
import sqlvalidator

from sqlparse.sql import Identifier, IdentifierList
from sqlparse.tokens import Keyword, DML, Whitespace

Evaluation/ 3 métriques:

valid_pred : True if te pred query is semanticaly correct, False otherwise. It is not really reliable.

keyword_score [0, 1]: Equivalent of F1 score for SQL keywords presence. The pred query must contains the keywords of the ground truth without adding new keywords.

identifier_score [0, 1]: For each keywords, f1 score is computed for identifiers words (table name, attributes...). The average gives the identifier_score.

In [11]:
def get_words_between_keywords(sql):
    parsed = sqlparse.parse(sql)[0]  # Parse the SQL statement
    tokens = parsed.tokens

    words_between = []  # To store words between keywords
    current_keywords = []
    buffer = []

    for token in tokens:
        if token.ttype in (Keyword, DML):  # Check if the token is a keyword
            if buffer:  # If there are words in the buffer, add them
                words_between.append((current_keywords[-1] if current_keywords else None, buffer))
                buffer = []
            current_keywords.append(token.value.upper())
        elif token.ttype is Whitespace:  # Ignore whitespace
            continue
        else:
            if isinstance(token, (Identifier, IdentifierList)):
                buffer.append(token.get_real_name() or token.value)
            else:
                buffer.append(token.value)

    if buffer:  # Add remaining buffer
        words_between.append((current_keywords[-1] if current_keywords else None, buffer))

    return words_between

In [103]:
def compute_f1score(a: set, b: set):
        """Equivalent of F1 score metric"""
        TP = len(a & b)
        FP = len(a - b)
        FN = len(b - a)

        if TP + 0.5*(FP + FN)==0:
          return 0

        return TP / (TP + 0.5*(FP + FN))

def compute_metrics(pred: str, ground_truth: str):


        # Check if the predicted query is semanticaly correct
  parsed_pred = sqlvalidator.parse(pred)
  valid_pred = False
  try:
      if parsed_pred.is_valid():
          valid_pred = True
  except:
      valid_pred = False

  # Normalize queries
  normalized_pred = sqlparse.format(pred, reindent=False, keyword_case='upper')
  normalized_gt = sqlparse.format(ground_truth, reindent=False, keyword_case='upper')

  # Compare Semantic of queries
  tokens_pred = get_words_between_keywords(normalized_pred)
  tokens_gt = get_words_between_keywords(normalized_gt)

    ## Keyword score

  keywords_pred = set([elem[0] for elem in tokens_pred])
  keywords_gt = set([elem[0] for elem in tokens_gt])

  keyword_score = compute_f1score(keywords_pred, keywords_gt)

  ## Identifier score

  identifier_score = 0
  commun_keywords = keywords_pred & keywords_gt
  for kw in list(commun_keywords):
      identifiers_pred = next((item for item in tokens_pred if item[0] == kw), None)[1]
      identifiers_gt = next((item for item in tokens_gt if item[0] == kw), None)[1]

      identifier_score += compute_f1score(set(identifiers_pred), set(identifiers_gt))

  if len(commun_keywords)==0:
    identifier_score=0
  else:
    identifier_score = identifier_score / len(commun_keywords)

  return valid_pred, keyword_score, identifier_score

Partie 3 : Zero shot prompt evaluation on 100 examples of the test dataset

In [173]:
base_prompt = """
You are an expert in SQL and database querying.
Always return **only** the SQL query, without any explanation, preamble, or additional text.
Your task is to convert natural language question after <<<>>> into optimized SQL queries.
<<<
Sentence: {sentence} Answer:
>>>
""".strip()


In [179]:
from tqdm import tqdm

predictions=[]
ground_truths=[]

for ex in tqdm(test):
  prompt = base_prompt.format(sentence=ex["question"])
  answer = generate(prompt)
  pred=extract_sql_answer(answer)
  predictions.append(pred)
  ground_truths.append(ex["SQL_query"])


100%|██████████| 100/100 [08:48<00:00,  5.28s/it]


In [180]:
print(predictions)
print(ground_truths)

['SELECT COUNT(*) FROM singer;', 'SELECT COUNT(*) FROM Singers', 'SELECT name, country, age FROM singers ORDER BY age DESC;', 'SELECT name, country, age FROM singer ORDER BY age DESC;', 'SELECT AVG(age), MIN(age), MAX(age) FROM singer', 'SELECT AVG(age), MIN(age), MAX(age) FROM singer', 'SELECT name, release_year FROM song WHERE singer_id = (SELECT singer', 'SELECT name, release_year FROM songs WHERE singer_id = (SELECT id FROM', 'SELECT DISTINCT country FROM singer WHERE age > 20;', 'SELECT DISTINCT country FROM singer WHERE age > 20;', 'SELECT country, COUNT(*) FROM singers GROUP BY country;', 'SELECT country, COUNT(*) FROM singer GROUP BY country;', 'SELECT s.name FROM singer s JOIN song s2 ON s.id', 'SELECT song_name FROM songs WHERE singer_id IN (     SELECT s', 'SELECT location, name FROM stadium WHERE capacity BETWEEN 5000 AND', 'SELECT name, location FROM station WHERE capacity BETWEEN 5000 AND 10', 'SELECT MAX(capacity), AVG(capacity) FROM stadium;', 'SELECT AVG(capacity), MAX

In [181]:
valid=0
keyword_score_add=0
identifier_score_add=0

for pred, gt in zip(predictions, ground_truths):
  valid_pred, keyword_score, identifier_score = compute_metrics(pred, gt)
  if valid_pred:
    valid+=1
  keyword_score_add+=keyword_score
  identifier_score_add+=identifier_score

valid_pourcent=(valid/len(predictions))*100
keyword_score_mean=keyword_score_add/len(predictions)
identifier_score_mean=identifier_score_add/len(predictions)

print(f' The valid prediction is :',{valid_pourcent}, '%')
print(f'keyword_score is ',{keyword_score_mean})
print(f'identifier_score is ',{identifier_score_mean})






81.0
0.7726269841269838
0.2338888888888888


For the zero shot prompt, the valid_pred is valid_pred : True if te pred query is semanticaly correct, False otherwise. It is not really reliable.

keyword_score [0, 1]: Equivalent of F1 score for SQL keywords presence. The pred query must contains the keywords of the ground truth without adding new keywords.

identifier_score [0, 1]: For each keywords, f1 score is computed for identifiers words (table name, attributes...). The average gives the identifier_score.

Granite Code 20b has a context window size of 8192, which means that we can use up to 8192 tokens to set the context. In our case, we can pass information about the database schema in order to get more accurate answers from the model.

Partie 4 : Few-shot example

In [182]:
base_fewshot_prompt = """
You are an expert in SQL and database querying.
Always return **only** the SQL query, without any explanation, preamble, or additional text.


Your task is to convert natural language question after <<<>>> into optimized SQL queries.


####
Here are some examples:

{examples}
####

<<<
Sentence: {sentence} Answer:
>>>
""".strip()

In [183]:
def format_demo(demo):
  string = demo['question']
  label = demo["SQL_query"]
  return "Sentence: " + string + " Answer : " + label

In [184]:
from tqdm import tqdm

demo=format_demo(test[0])
predictions=[]
ground_truths=[]

for ex in tqdm(test):
  prompt = base_fewshot_prompt.format(examples=demo,sentence=ex["question"])
  answer = generate(prompt)
  pred=extract_sql_answer(answer)
  predictions.append(pred)
  ground_truths.append(ex["SQL_query"])


100%|██████████| 100/100 [09:05<00:00,  5.45s/it]


In [185]:
print(predictions)
print(ground_truths)

['SELECT count(*) FROM singer', 'SELECT count(*) FROM singer', 'SELECT name, country, age FROM singer ORDER BY age DESC;', 'SELECT name, country, age FROM singer ORDER BY age DESC', 'SELECT AVG(age), MIN(age), MAX(age) FROM singer', 'SELECT AVG(age), MIN(age), MAX(age) FROM singer', 'SELECT singer.name, song.release_year FROM singer JOIN', 'SELECT singer.name, song.release_year FROM singer, song WHERE s', 'SELECT DISTINCT country FROM singer WHERE age > 20;', 'SELECT country FROM singer WHERE age > 20 GROUP BY country', 'SELECT country, count(*) FROM singer GROUP BY country', 'SELECT country, count(*) FROM singer GROUP BY country', 'SELECT singer.name, song.title FROM singer JOIN song ON', 'SELECT singer.name, song.title FROM singer JOIN song ON', 'SELECT location, name FROM stadium WHERE capacity BETWEEN 5000 AND 1', 'SELECT location, name FROM station WHERE capacity BETWEEN 5000 AND 10', 'SELECT max(capacity), avg(capacity) FROM stadium', 'SELECT AVG(capacity), MAX(capacity) FROM sta

In [186]:
valid=0
keyword_score_add=0
identifier_score_add=0

for pred, gt in zip(predictions, ground_truths):
  valid_pred, keyword_score, identifier_score = compute_metrics(pred, gt)
  if valid_pred:
    valid+=1
  keyword_score_add+=keyword_score
  identifier_score_add+=identifier_score

valid_pourcent=(valid/len(predictions))*100
keyword_score_mean=keyword_score_add/len(predictions)
identifier_score_mean=identifier_score_add/len(predictions)

print(valid_pourcent)
print(keyword_score_mean)
print(identifier_score_mean)

78.0
0.7644126984126979
0.3347222222222222
